### PR TITLE
feat(cli): declarative deployment

### PR DESCRIPTION
**Pull request by holic**
- Deploy step now takes in a fully-resolved config and an optional world address, inspects the world, and does the minimal set of operations necessary to bring the world in alignment with the config.
  - This deploy function attempts to look at the world from the deploy block number to the block number when the function was called, to keep all the world values in alignment (since they may be reading from the chain at different times/blocks during the deploy process). This may cause issues with certain RPCs that limit block ranges on read queries.
  - This isn't "fully" declarative and could later be improved by separating the "plan" step from the "apply" step (ala terraform), but I didn't want to spend too much time on this. The bones are there to move in this direction though!
  - This doesn't have a "generate MUD config from world" yet, but again, the bones are there to build off of.
  - ~~This doesn't check namespace ownership/access yet and will throw an error or revert if we attempt an operation on a namespace we don't have access to. This could be improved by doing some checks during the "plan" step.~~ This is now included.
- `dev-contracts` CLI command is simplified to 1) watch files that would affect the world and 2) rebuild/redeploy everything. Since the deploy step is much smarter and only does the minimal work necessary, I am hoping the speed gains here can justify re-running build for now (to keep the logic simpler). Worth a refactor in the future when we have more time to spend on making this great.

### CODE REVIEW

File: examples/minimal/packages/contracts/script/PostDeploy.s.sol
Comment (holic) [inline examples/minimal/packages/contracts/script/PostDeploy.s.sol:31]: these operations are not idempotent so I am not sure if we should keep + adapt or remove

File: packages/cli/src/commands/dev-contracts.ts
Comment (holic) [inline packages/cli/src/commands/dev-contracts.ts:76]: we now run this in its entirety ~every time, rather than splitting up just codegen vs forge build etc. both for ease and for safety

I want this to force us to reevaluate when to run these things and the downstream effects of e.g. not running and potentially optimize in different ways (e.g. make declarative deployment smarter about the work it does rather than not running it at all)

File: packages/cli/src/deploy/logsToWorldDeploy.ts
Comment (alvrs) [inline packages/cli/src/deploy/logsToWorldDeploy.ts:27]: Imo we can just assume that, since the verified contracts only emit it once, and if a user actively re-emits it, it's either their fault if it breaks, or maybe it's for some intentional advanced use case. (There are many other ways in which users could deliberately make things break in their own custom World that we don't validate either.)

File: packages/cli/src/deploy/resolveConfig.ts
Comment (alvrs) [inline packages/cli/src/deploy/resolveConfig.ts:92]:
```suggestion
  // ugh (https://github.com/latticexyz/mud/issues/1668)
```

File: packages/cli/src/index.ts
Comment (alvrs) [inline packages/cli/src/index.ts:1]: do we need this file?

File: packages/cli/src/runDeploy.ts
Comment (alvrs) [inline packages/cli/src/runDeploy.ts:38]:
```suggestion
/**
 * Given some CLI arguments, finds and resolves a MUD config, foundry profile, and runs a deploy.
 * This is used by the deploy, test, and dev-contracts CLI commands.
 */
```

File: packages/cli/src/runDeploy.ts
Comment (alvrs) [inline packages/cli/src/runDeploy.ts:111]: nit (trying to avoid horizontal scroll bars and better vim support)
```suggestion
      // We expect the worlds file to be committed and since local deployments are often
      // a consistent address but different block number, we'll ignore the block number.
```

File: packages/cli/src/runDeploy.ts
Comment (holic) [reply packages/cli/src/runDeploy.ts:1]: I've kept everything in `src/deploy` unaware of all the CLI arg parsing, etc., because it felt like all of that could use a rethink. Wanted to start/keep a "clean slate". I don't even love that `resolveConfig` is in there, but hoping to just completely delete it once the new config lands.

File: examples/minimal/packages/contracts/script/PostDeploy.s.sol
Comment (holic) [reply examples/minimal/packages/contracts/script/PostDeploy.s.sol:31]: sounds good, I would prefer this too!

if we want "post upgrade" scripts/hooks, we can introduce a `PostUpgrade.s.sol`

File: packages/cli/src/commands/dev-contracts.ts
Comment (holic) [reply packages/cli/src/commands/dev-contracts.ts:50]: It is!

I think I may have confused this line with the test runner, but I'd like to figure out how things can exit cleanly rather than us having to manually `process.exit(0)` everywhere. Not sure if this is a yargs thing, or a long running anvil node thing, or what.

File: packages/cli/src/commands/dev-contracts.ts
Comment (holic) [reply packages/cli/src/commands/dev-contracts.ts:74]: oh yeah, I can remove the debounce now, that was before I started fiddling with `exhaustMap` approach

File: packages/cli/src/commands/dev-contracts.ts
Comment (holic) [reply packages/cli/src/commands/dev-contracts.ts:75]: right now things just explode and I tried to wrap in a try/catch but still got weird behavior

ideally things would re-run once files change, regardless of error, but didn't wanna rabbit hole too long on that

File: packages/cli/src/deploy/deploy.ts
Comment (holic) [reply packages/cli/src/deploy/deploy.ts:18]: surely vim can wrap lines if you ask it to?

we should add a linter/autoformatter if this is what we want everywhere, so that we're wrapping consistently

File: packages/cli/src/deploy/deploy.ts
Comment (holic) [reply packages/cli/src/deploy/deploy.ts:61]: the thing is we're waiting for ~dozens of txs here, any of which can revert

which do we throw with?

File: packages/cli/src/commands/dev-contracts.ts
Comment (holic) [reply packages/cli/src/commands/dev-contracts.ts:73]: actually, debounce is still useful here in cases where there are a bunch of simultaneous file updates, so we can try to wait for the ~last update in a batch rather than the first one seen, gonna leave this and even up the timeout a bit

File: packages/cli/src/deploy/common.ts
Comment (holic) [reply packages/cli/src/deploy/common.ts]: clarified to 
> Block number at the time of fetching world deploy. We use this block number when requesting data from the chain to align chain state with the same block during the introspection steps of the deploy.

but now wondering it's worth renaming the variables 
- `fromBlock` -> `deployBlock`
- `toBlock` -> `stateBlock` or `snapshotBlock` or something

File: packages/cli/src/deploy/deployWorld.ts
Comment (holic) [reply packages/cli/src/deploy/deployWorld.ts:28]: https://github.com/wagmi-dev/viem/pull/1330

File: packages/cli/src/deploy/getFunctions.ts
Comment (holic) [reply packages/cli/src/deploy/getFunctions.ts:17]: Since we don't have any sort of "index" for tables like FunctionSelectors, we look at all the keys used in the logs for the table. But since viem can only filter on one event+args (could be an RPC node limitation), we can't easily fetch all store logs for a given table ID and recompute the current state of the table. So we just use the one event that is actually used internally.

File: packages/cli/src/deploy/common.ts
Comment (holic) [reply packages/cli/src/deploy/common.ts]: renamed to deployBlock + stateBlock

File: packages/cli/src/deploy/deployWorld.ts
Comment (holic) [reply packages/cli/src/deploy/deployWorld.ts]: does this clarify it? https://github.com/latticexyz/mud/pull/1702#discussion_r1353038491

File: packages/cli/src/deploy/ensureFunctions.ts
Comment (holic) [reply packages/cli/src/deploy/ensureFunctions.ts:31]: I didn't want to throw here and stop mid-deploy. I agree that it would be better to throw, but maybe after we have a separation of "plan" and "apply" steps?

File: packages/cli/src/deploy/getResourceAccess.ts
Comment (holic) [reply packages/cli/src/deploy/getResourceAccess.ts:46]: weird, I don't see it:
image

File: packages/cli/src/index.ts
Comment (holic) [reply packages/cli/src/index.ts:1]: unclear! we have this pattern in store-indexer too and I can't recall why I ended up with that, maybe just to preserve the scaffolding necessary to start exporting from this package

File: packages/cli/src/deploy/getFunctions.ts
Comment (holic) [reply packages/cli/src/deploy/getFunctions.ts:30]: not seeing this!

File: packages/cli/src/deploy/getTables.ts
Comment (holic) [reply packages/cli/src/deploy/getTables.ts:39]: there are some small differences and it felt weird to have a dependency on store-sync here, so I wanted to think more about where to put it (e.g. common or somewhere else)

File: packages/cli/src/deploy/getSystems.ts
Comment (holic) [reply packages/cli/src/deploy/getSystems.ts]: whoops, forgot to finish these (ended up not being used anyway)

File: packages/cli/src/commands/dev-contracts.ts
Comment (alvrs) [reply packages/cli/src/commands/dev-contracts.ts:75]: Yeah if i'm not mistaken that was the previous behavior (e.g. if there is a typo in a contract the dev runner logged the error but kept watching to rerun when the typo was fixed). I'm fine with not blocking this PR on this, but think we should do it in a followup soon.

Comment (alvrs) [reply packages/cli/src/commands/dev-contracts.ts:75]: added a follow up issue here: https://github.com/latticexyz/mud/issues/1731

File: packages/cli/src/deploy/deploy.ts
Comment (alvrs) [reply packages/cli/src/deploy/deploy.ts:81]: mhm i see - the implementation here just reverts with the first error (right?), that seems fine to me

File: packages/cli/src/deploy/deployWorld.ts
Comment (alvrs) [reply packages/cli/src/deploy/deployWorld.ts]: yep much clearer!

File: packages/cli/src/deploy/getFunctions.ts
Comment (alvrs) [reply packages/cli/src/deploy/getFunctions.ts:17]: got it, thanks for the context

File: packages/cli/src/deploy/ensureSystems.ts
Comment (holic) [inline packages/cli/src/deploy/ensureSystems.ts:45]: cc @CrazyNorman

there's a few other TODOs for `batchCall` in here but I think this is one place we can add it that has meaningful impact to the deploy functionality vs. just making it faster

Comment (CrazyNorman) [reply packages/cli/src/deploy/ensureSystems.ts:45]: > cc @CrazyNorman
> 
> there's a few other TODOs for `batchCall` in here but I think this is one place we can add it that has meaningful impact to the deploy functionality vs. just making it faster

Got you. So If I'm getting you right, we can use `batchCall` to do atomic setup, handling each system as a whole, one by one. This will make the deployment steps more modular.

**Review by alvrs — APPROVED** [review]
i'm very excited about this! Let's 🚢

File: packages/cli/src/deploy/ensureSystems.ts
Comment (holic) [reply packages/cli/src/deploy/ensureSystems.ts:45]: Exactly! If we for some reason fail to register a system but add access control, or vice versa, that feels like a weird state to be in. Using `batchCall` here should help keep the operation atomic so if one step in setting up the system fails, the entire system setup fails.

**Pull request merged**
